marimo/_output/formatters/arviz_formatters.py

```diff
@@ -0,0 +1,120 @@
+from __future__ import annotations
+
+from typing import Any, Union
+from marimo._messaging.mimetypes import KnownMimeType
+from marimo._output.formatters.formatter_factory import FormatterFactory
+import matplotlib.pyplot as plt
+import numpy as np
+
+class ArviZFormatter(FormatterFactory):
+    @staticmethod
+    def package_name() -> str:
+        return "arviz"
+
+    def register(self) -> None:
+        import arviz as az
+        from marimo._output import formatting
+
+        @formatting.formatter(az.InferenceData)
+        def _format_inference_data(
+            data: az.InferenceData,
+        ) -> tuple[KnownMimeType, str]:
+            return ("text/plain", str(data))
+
+        @formatting.formatter(np.ndarray)
+        def _format_ndarray(
+            arr: np.ndarray,
+        ) -> tuple[KnownMimeType, str]:
+            return self.format_numpy_axes(arr)
+
+        @formatting.formatter(dict)
+        def _format_dict(
+            d: dict,
+        ) -> tuple[KnownMimeType, str]:
+            return self.format_dict_with_plot(d)
+
+        @formatting.formatter(plt.Figure)
+        def _format_figure(
+            fig: plt.Figure,
+        ) -> tuple[KnownMimeType, str]:
+            return self.format_figure(fig)
+
+        @formatting.formatter(object)
+        def _format_arviz_plot(
+            obj: Any,
+        ) -> tuple[KnownMimeType, str]:
+            return self.format_arviz_plot(obj)
+
+    @classmethod
+    def format_numpy_axes(cls, arr: np.ndarray) -> tuple[KnownMimeType, str]:
+        # Check if array contains axes (to render plots) or not
+        if arr.dtype == object and cls._contains_axes(arr):
+            fig = plt.gcf()
+            if fig.get_axes():  # Only process if there are axes to show
+                axes_info = cls._get_axes_info(fig)
+                plot_html = cls._get_plot_html(fig)
+                plt.close(fig)  # Safely close the figure after saving
+                combined_html = f"<pre>{axes_info}</pre><br>{plot_html}"
+                return ("text/html", combined_html)
+        # Fallback to plain text if no axes or plot are present
+        return ("text/plain", str(arr))
+
+    @staticmethod
+    def _contains_axes(arr: np.ndarray) -> bool:
+        if arr.ndim == 1:
+            return any(isinstance(item, plt.Axes) for item in arr)
```

Comment:
should we put a cap at the length of `arr`? this could be a array with 1 million items but dimension 1 and it will still run.

i think you can put a small cap. would you mind adding a comment too?

Reply:
Did length handling as suggested. Added comments for the same.

Sample code:
```python
@staticmethod
    def _contains_axes(arr: np.ndarray) -> bool:
        """
        Check if the numpy array contains any matplotlib Axes objects.
        To ensure performance for large arrays, we limit the check to the first 1000 items.
        This should be sufficient for most use cases while avoiding excessive computation time.
        """
        # Cap the number of items to check for performance reasons
        MAX_ITEMS_TO_CHECK = 1000

        if arr.ndim == 1:
            # For 1D arrays, check up to MAX_ITEMS_TO_CHECK items
            return any(isinstance(item, plt.Axes) for item in arr[:MAX_ITEMS_TO_CHECK])
        elif arr.ndim == 2:
            # For 2D arrays, check up to MAX_ITEMS_TO_CHECK items in total
            items_checked = 0
            for row in arr:
                for item in row:
                    if isinstance(item, plt.Axes):
                        return True
                    items_checked += 1
                    if items_checked >= MAX_ITEMS_TO_CHECK:
                        return False
        return False
```

Reply:
should it be all() instead of any()?

Reply:
also are there real examples with 1000 axis?

Reply:
> should it be all() instead of any()?

I feel that the use of `any()` is right here. We're checking if there's at least one Axes object in the array, not if all items are Axes objects. `any()` returns True as soon as it finds a matching item, which is what I believe would be ideal to ArviZ related cases.

ArviZ related plot information which I tested - https://github.com/marimo-team/marimo/pull/1668#issuecomment-2219743427

Reply:
> also are there real examples with 1000 axis?

Typical use cases rarely exceed a few dozen axes at most. We can significantly reduce this number to maybe 100 or even 50?

Reply:
Yea reducing to 100 or failing faster is more ideal. Otherwise, this will loop through 1000 items for every non Axis np array which can be wasteful

Reply:
> Yea reducing to 100 or failing faster is more ideal. Otherwise, this will loop through 1000 items for every non Axis np array which can be wasteful

Modified it now.